File: sdks/redlite-java/src/main/java/com/redlite/namespaces/HistoryNamespace.java
package com.redlite.namespaces;

import com.redlite.Redlite;
import com.redlite.RedliteException;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * Version history namespace for Redlite HISTORY.* commands.
 */
public class HistoryNamespace {
    private final Redlite client;

    public HistoryNamespace(Redlite client) {
        this.client = client;
    }

    // =========================================================================
    // History Enable/Disable Commands
    // =========================================================================

    /**
     * Enable history tracking globally.
     *
     * @param retentionType  "unlimited", "time", or "count"
     * @param retentionValue Value for time (ms) or count retention
     */
    public void enableGlobal(String retentionType, long retentionValue) {
        throw new RedliteException("History commands not yet implemented");
    }

    /**
     * Enable history tracking globally with unlimited retention.
     */
    public void enableGlobal() {
        enableGlobal("unlimited", 0);
    }

    /**
     * Enable history tracking for a specific database.
     *
     * @param dbNum          Database number
     * @param retentionType  "unlimited", "time", or "count"
     * @param retentionValue Value for time (ms) or count retention
     */
    public void enableDatabase(int dbNum, String retentionType, long retentionValue) {
        throw new RedliteException("History commands not yet implemented");
    }

    /**
     * Enable history tracking for a specific database with unlimited retention.
     */
    public void enableDatabase(int dbNum) {
        enableDatabase(dbNum, "unlimited", 0);
    }

    /**
     * Enable history tracking for a specific key.
     *
     * @param key            Key to enable history for
     * @param retentionType  "unlimited", "time", or "count"
     * @param retentionValue Value for time (ms) or count retention
     */
    public void enableKey(String key, String retentionType, long retentionValue) {
        throw new RedliteException("History commands not yet implemented");
    }

    /**
     * Enable history tracking for a specific key with unlimited retention.
     */
    public void enableKey(String key) {
        enableKey(key, "unlimited", 0);
    }

    /**
     * Disable history tracking globally.
     */
    public void disableGlobal() {
        throw new RedliteException("History commands not yet implemented");
    }

    /**
     * Disable history tracking for a specific database.
     *
     * @param dbNum Database number
     */
    public void disableDatabase(int dbNum) {
        throw new RedliteException("History commands not yet implemented");
    }

    /**
     * Disable history tracking for a specific key.
     *
     * @param key Key to disable history for
     */
    public void disableKey(String key) {
        throw new RedliteException("History commands not yet implemented");
    }

    /**
     * Check if history tracking is enabled for a key.
     *
     * @param key Key to check
     * @return true if history is enabled
     */
    public boolean isEnabled(String key) {
        throw new RedliteException("History commands not yet implemented");
    }

    // =========================================================================
    // Legacy API (deprecated, use enable/disable methods above)
    // =========================================================================

    /**
     * Enable history tracking for a key pattern.
     *
     * @deprecated Use enableGlobal(), enableDatabase(), or enableKey() instead
     */
    @Deprecated
    public boolean enable(String pattern, int maxVersions) {
        throw new RedliteException("History commands not yet implemented");
    }

    /**
     * Enable history tracking for all keys.
     *
     * @deprecated Use enableGlobal() instead
     */
    @Deprecated
    public boolean enable() {
        enableGlobal();
        return true;
    }

    /**
     * Disable history tracking for a key pattern.
     *
     * @deprecated Use disableGlobal(), disableDatabase(), or disableKey() instead
     */
    @Deprecated
    public boolean disable(String pattern) {
        throw new RedliteException("History commands not yet implemented");
    }

    /**
     * Get version history of a key.
     */
    public List<HistoryEntry> get(String key) {
        return get(key, 0, -1);
    }

    /**
     * Get version history of a key with range.
     */
    public List<HistoryEntry> get(String key, long start, long end) {
        // Would execute HISTORY.GET command
        return Collections.emptyList();
    }

    /**
     * Get a specific version of a key.
     */
    public @Nullable byte[] getVersion(String key, long version) {
        // Would execute HISTORY.GETVERSION command
        return null;
    }

    /**
     * Get the number of versions stored for a key.
     */
    public long count(String key) {
        // Would execute HISTORY.COUNT command
        return 0;
    }

    /**
     * Revert a key to a previous version.
     */
    public boolean revert(String key, long version) {
        // Would execute HISTORY.REVERT command
        return true;
    }

    /**
     * Trim old versions of a key.
     */
    public long trim(String key, int keepVersions) {
        // Would execute HISTORY.TRIM command
        return 0;
    }

    /**
     * Get history information for a key.
     */
    public Map<String, Object> info(String key) {
        // Would execute HISTORY.INFO command
        return Collections.emptyMap();
    }

    /**
     * A single history entry for a key.
     */
    public record HistoryEntry(long version, long timestamp, @Nullable byte[] value) {
    }
}

File: sdks/redlite-kotlin/src/main/kotlin/com/redlite/namespaces/HistoryNamespace.kt
package com.redlite.namespaces

import com.redlite.Redlite
import com.redlite.RedliteException

/**
 * Version history namespace for Redlite HISTORY.* commands.
 */
class HistoryNamespace(private val client: Redlite) {

    // =========================================================================
    // History Enable/Disable Commands
    // =========================================================================

    /**
     * Enable history tracking globally.
     *
     * @param retentionType "unlimited", "time", or "count"
     * @param retentionValue Value for time (ms) or count retention
     */
    fun enableGlobal(retentionType: String = "unlimited", retentionValue: Long = 0) {
        throw RedliteException("History commands not yet implemented")
    }

    /**
     * Enable history tracking for a specific database.
     *
     * @param dbNum Database number
     * @param retentionType "unlimited", "time", or "count"
     * @param retentionValue Value for time (ms) or count retention
     */
    fun enableDatabase(dbNum: Int, retentionType: String = "unlimited", retentionValue: Long = 0) {
        throw RedliteException("History commands not yet implemented")
    }

    /**
     * Enable history tracking for a specific key.
     *
     * @param key Key to enable history for
     * @param retentionType "unlimited", "time", or "count"
     * @param retentionValue Value for time (ms) or count retention
     */
    fun enableKey(key: String, retentionType: String = "unlimited", retentionValue: Long = 0) {
        throw RedliteException("History commands not yet implemented")
    }

    /**
     * Disable history tracking globally.
     */
    fun disableGlobal() {
        throw RedliteException("History commands not yet implemented")
    }

    /**
     * Disable history tracking for a specific database.
     *
     * @param dbNum Database number
     */
    fun disableDatabase(dbNum: Int) {
        throw RedliteException("History commands not yet implemented")
    }

    /**
     * Disable history tracking for a specific key.
     *
     * @param key Key to disable history for
     */
    fun disableKey(key: String) {
        throw RedliteException("History commands not yet implemented")
    }

    /**
     * Check if history tracking is enabled for a key.
     *
     * @param key Key to check
     * @return true if history is enabled
     */
    fun isEnabled(key: String): Boolean {
        throw RedliteException("History commands not yet implemented")
    }

    // =========================================================================
    // Legacy API (deprecated, use enable/disable methods above)
    // =========================================================================

    /**
     * Enable history tracking for a key pattern.
     *
     * @param pattern Key pattern (e.g., "user:*")
     * @param maxVersions Maximum versions to keep per key
     * @deprecated Use enableGlobal(), enableDatabase(), or enableKey() instead
     */
    @Deprecated("Use enableGlobal(), enableDatabase(), or enableKey() instead")
    fun enable(pattern: String = "*", maxVersions: Int = 100): Boolean {
        client.execute("HISTORY.ENABLE", pattern, maxVersions.toString())
        return true
    }

    /**
     * Disable history tracking for a key pattern.
     *
     * @deprecated Use disableGlobal(), disableDatabase(), or disableKey() instead
     */
    @Deprecated("Use disableGlobal(), disableDatabase(), or disableKey() instead")
    fun disable(pattern: String = "*"): Boolean {
        client.execute("HISTORY.DISABLE", pattern)
        return true
    }

    // =========================================================================
    // History Query Commands
    // =========================================================================

    /**
     * Get version history of a key.
     *
     * @param key The key
     * @param start Start version (0 = oldest)
     * @param end End version (-1 = newest)
     */
    fun get(key: String, start: Long = 0, end: Long = -1): List<HistoryEntry> {
        val result = client.execute("HISTORY.GET", key, start.toString(), end.toString())
        if (result is List<*>) {
            return result.mapNotNull { entry ->
                if (entry is List<*> && entry.size >= 3) {
                    HistoryEntry(
                        version = (entry[0] as? Number)?.toLong() ?: 0,
                        timestamp = (entry[1] as? Number)?.toLong() ?: 0,
                        value = entry[2] as? ByteArray
                    )
                } else null
            }
        }
        return emptyList()
    }

    /**
     * Get a specific version of a key.
     */
    fun getVersion(key: String, version: Long): ByteArray? {
        val result = client.execute("HISTORY.GETVERSION", key, version.toString())
        return result as? ByteArray
    }

    /**
     * Get the number of versions stored for a key.
     */
    fun count(key: String): Long {
        val result = client.execute("HISTORY.COUNT", key)
        return (result as? Number)?.toLong() ?: 0
    }

    /**
     * Revert a key to a previous version.
     */
    fun revert(key: String, version: Long): Boolean {
        client.execute("HISTORY.REVERT", key, version.toString())
        return true
    }

    /**
     * Trim old versions of a key.
     */
    fun trim(key: String, keepVersions: Int): Long {
        val result = client.execute("HISTORY.TRIM", key, keepVersions.toString())
        return (result as? Number)?.toLong() ?: 0
    }

    /**
     * Get history information for a key.
     */
    @Suppress("UNCHECKED_CAST")
    fun info(key: String): Map<String, Any> {
        val result = client.execute("HISTORY.INFO", key)
        if (result is List<*>) {
            val map = mutableMapOf<String, Any>()
            var i = 0
            while (i < result.size - 1) {
                val k = result[i]?.toString() ?: continue
                val v = result[i + 1] ?: continue
                map[k] = v
                i += 2
            }
            return map
        }
        return emptyMap()
    }
}

/**
 * A single history entry for a key.
 */
data class HistoryEntry(
    val version: Long,
    val timestamp: Long,
    val value: ByteArray?
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as HistoryEntry
        if (version != other.version) return false
        if (timestamp != other.timestamp) return false
        if (!value.contentEquals(other.value)) return false
        return true
    }

    override fun hashCode(): Int {
        var result = version.hashCode()
        result = 31 * result + timestamp.hashCode()
        result = 31 * result + (value?.contentHashCode() ?: 0)
        return result
    }
}
